src/parser/shared/normalizers/Channeling.ts

```diff
@@ -0,0 +1,344 @@
+import SPELLS from 'common/SPELLS';
+import CASTS_THAT_ARENT_CASTS from 'parser/core/CASTS_THAT_ARENT_CASTS';
+import {
+  AnyEvent,
+  BeginCastEvent,
+  BeginChannelEvent,
+  CastEvent,
+  EndChannelEvent,
+  EventType,
+  HasAbility,
+  HasSource,
+} from 'parser/core/Events';
+import EventsNormalizer from 'parser/core/EventsNormalizer';
+import InsertableEventsWrapper from 'parser/core/InsertableEventsWrapper';
+import { Options } from 'parser/core/Module';
+
+/**
+ * Channels and casts are handled differently in events, and some information is also missing and must be inferred.
+ * This normalizer fabricates new BeginChannel and EndChannel events that delineate the cast times
+ * for both casts and channels, and additionally includes inferred information.
+ *
+ * Instant spells have only a Cast event - they are ignored by this normalizer, which
+ * is only concerned with cast time abilities.
+ *
+ * Spells with a cast time have a BeginCast event when they are started and and a Cast event when
+ * they finish. If they are cancelled before finishing, there will be a BeginCast without the linked
+ * Cast. For completed casts, this normalizer fabricates a BeginChannel for the start of the spell
+ * and an EndChannel for the finish. For cancelled casts, there will only be a BeginChannel event
+ * with the 'isCancelled' flag set.
+ *
+ * Channelled spells are handled inconsistently in the events, but this normalizer aims to uniformly
+ * represent them with a BeginChannel when started and and EndChannel when stopped (regardless of if
+ * the spell finished or it was stopped early). There are a variety of ways channeled spells show
+ * in events, and this normalizer allows special case handling to be registered for each.
+ */
+class Channeling extends EventsNormalizer {
+  /**
+   * Listing of all special case handlers for channels
+   */
+  static CHANNEL_SPECS: ChannelSpec[] = [
+    // General
+    // Shadowlands Encounter
+    buffChannelSpec(SPELLS.SOUL_INFUSION.id), // fight channel from Sun King's Salvation - see in this log: https://wowanalyzer.com/report/g4Pja6pLHnmQtbvk/32-Normal+Sun+King's+Salvation+-+Kill+(10:14)/Pjurbo/standard/events
+    // Mage
+    buffChannelSpec(SPELLS.EVOCATION.id),
+    buffChannelSpec(SPELLS.SHIFTING_POWER.id),
+    nextCastChannelSpec(SPELLS.ARCANE_MISSILES.id),
+    // Warlock
+    buffChannelSpec(SPELLS.DRAIN_SOUL_TALENT.id),
+    // Priest
+    buffChannelSpec(SPELLS.DIVINE_HYMN_CAST.id),
+    nextCastChannelSpec(SPELLS.PENANCE_CAST.id),
+    buffChannelSpec(SPELLS.VOID_TORRENT_TALENT.id),
+    buffChannelSpec(SPELLS.MIND_FLAY.id), // TODO double check ID
+    buffChannelSpec(SPELLS.MIND_SEAR.id), // TODO double check ID
+    // Rogue
+    // Druid
+    buffChannelSpec(SPELLS.CONVOKE_SPIRITS.id),
+    // Monk
```

Comment:
Does this work with pets too? 
Invoke Yu'lon, the Jade Serpent currently does all of **HER** healing though Soothing Breath which is a 4 second channel
Fallen Order currently channels Soothing Mist OR casts Enveloping Mist

Reply:
As currently set up it does not, but I think it would be possible to make it handle pets if I extended `ChannelState` to track per entity. It's generally important to filter by source because if you don't, you get screwed up by Cast events that target the player.

Is there value and supporting pets though? The primary purpose of the channel events is to support ActiveTime calculations, and I thought that wasn't really relevant for pets - I could be mistaken though as I've never played a pet class.